test/cpp_functor.cxx

```diff
@@ -0,0 +1,209 @@
+    QuadraticForm() = delete;
+    QuadraticForm(Vector A_, Vector b_) :
+      A(std::move(A_)),
+      b(std::move(b_))
+    {
+      if (A.size() != b.size() * b.size()) {
+        throw std::runtime_error(
+            "[QuadraticForm] matrix and bias dimension mismatch: " +
+            std::to_string(A.size()) +
+            " != " +
+            std::to_string(b.size() * b.size()));
+      }
+      dimension = b.size();
+    }
+
+    double compute_form(const double* x) const
+    {
+      // ½ xᵀAx + bᵀx
+      return compute_quadratic_term(x) + compute_linear_term(x);
+    }
+
+    void compute_gradient_in_place(const double* x, double* grad_array) const
+    {
+      // matrix is assumed symmetric, hence gradient is (Ax + b)
+      Vector grad_vec = compute_A_times_x(x);
+
+      auto b_iterator = std::begin(b);
+      for (auto it = std::begin(grad_vec); it != std::end(grad_vec); ) {
+        *it++ += *b_iterator++;
+      }
+
+      std::copy(std::begin(grad_vec), std::end(grad_vec), grad_array);
+    }
+
+}; // QuadraticForm
+
+
+class LinearRegression {
+  private:
+    QuadraticForm quadratic_form;
+
+  public:
+    LinearRegression() = delete;
+    LinearRegression(QuadraticForm quadratic_form_) :
+      quadratic_form(std::move(quadratic_form_)) {}
+
+    double operator()(unsigned n, const double* x, double* grad) const
+    {
+      const double result = quadratic_form.compute_form(x);
+      if (!!grad) {
+        quadratic_form.compute_gradient_in_place(x, grad);
+      }
+
+      return result;
+    }
+}; // LinearRegression
+
+
+class SineRegression {
+  private:
+    QuadraticForm quadratic_form;
+
+  public:
+    SineRegression() = delete;
+    SineRegression(QuadraticForm quadratic_form_) :
+      quadratic_form(std::move(quadratic_form_)) {}
+
+    double operator()(unsigned n, const double* x, double* grad) const
+    {
+      const double form_result = quadratic_form.compute_form(x);
+      const double result = sin(form_result);
+
+      if (!!grad) {
+        const double grad_coefficient = cos(form_result);
+        quadratic_form.compute_gradient_in_place(x, grad);
+
+        for (size_t i = 0; i < n; ++i) {
+          grad[i] *= grad_coefficient;
+        }
+      }
+
+      return result;
+    }
+}; // SineRegression
+
+
+
+int main()
+{
+  // quadratic form is positive-definite for tau ∈ [-1, 2]
+  constexpr double tau = 1.5;
+  constexpr size_t dimension = 3;
+
+  Vector A = {
+      2,   -1,   tau,
+     -1,    2,    -1,
+    tau,   -1,     2 };
+  Vector b = { -3, 8, 1 };
+
+  assert(b.size() == dimension);
+  assert(A.size() == dimension * dimension);
+
+  QuadraticForm    form(std::move(A), std::move(b));
+  LinearRegression objective_1(form);
+  SineRegression   objective_2(std::move(form));
+
+  nlopt::opt optimizer("LD_MMA", dimension);
+  optimizer.set_xtol_rel(1e-4);
+  optimizer.set_maxeval(1000);
+
+
+  // linear regression optimization
+  optimizer.set_min_objective(std::move(objective_1));
```

Comment:
My understanding is that `std::move` leaves the source object (`objective_1`) in a "valid but undefined state", which seems like not the sort of thing one would always want to do — what if the caller wants to use the objective function for something other than optimization?

In general, we should be clear about the resource management here.  Who is responsible for freeing the resources associated with the objective functor?  Do they get freed when the functor leaves the scope where it was defined, or do they get freed when the `optimizer` object is destroyed?

Reply:
In this example, I simply used `std::move` just because I myself don't need the `objective_1` afterwards in the `main` function, so it's better to move it rather than copy. But in practice, users should decide for themselves what they want: to make a copy or move the object. Since `set_min_objective` accepts the parameter by value, it will make a copy of the argument unless it's passed using `std::move` (or is an otherwise RHS expression, like nameless lambda). OK, this is a slight simplification because in reality `std::function`'s constructor is called in between, but the copy-or-move logic is still intact -- because `std::function` has a move ctor.

In general, since `set_min_objective`, `set_max_objective` accept a functor by value, it is assumed that `optimizer` is responsible for memory management of its own copy of the functor, but the outer one should be managed by the user. I believe this is fairly common, and actually the `std::move` allows the user to decide whether s/he wants to keep the object or pass it altogether, instead of hardcoding this into a library. Again, in this example I chose to "move" the object because I just don't need it afterwards. But we can also write simply
```cpp
optimizer.set_min_objective(objective_1);
```
and then the `objective_1` is usable (and a copy is made).

As for the inner workings, since functor is an STL-container, functor will get destroyed when `std::function`'s destructor is called. Which, in turn, happens when `myfunc_data` is freed. And that, in turn, is already well-designed within the library.

Do you agree with this approach?

Reply:
That seems fine to me.  We shouldn't use `std::move` in the documentation examples, however, because people are likely to copy that blindly and will then get into trouble.

Reply:
oh, okay! I could change that if you want? in a new PR? and thank you for merging!

Reply:
A new PR to clarify this would be fine.